src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,862 @@
+        /// 
+        /// ---------------------------------------------------->
+        /// 
+        /// public int M_f(int x, int y)
+        /// {
+        ///     return x * y;
+        /// }
+        /// 
+        /// public void M(int x, int y)
+        /// {
+        ///     int f = x * y;
+        /// }
+        /// </summary>
+        private static async Task<SyntaxNode> ExtractMethodAsync(Document document, TExpressionSyntax expression,
+            IMethodSymbol methodSymbol, ImmutableArray<IParameterSymbol> validParameters, string newMethodIdentifier, SyntaxGenerator generator, CancellationToken cancellationToken)
+        {
+
+            // Remove trailing trivia because it adds spaces to the beginning of the following statement.
+            var newStatements = ImmutableArray.CreateBuilder<SyntaxNode>();
+            newStatements.Add(generator.ReturnStatement(expression.WithoutTrailingTrivia()));
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var typeSymbol = semanticModel.GetTypeInfo(expression, cancellationToken).ConvertedType ?? semanticModel.Compilation.ObjectType;
+            var codeGenerationService = document.GetRequiredLanguageService<ICodeGenerationService>();
+            var newMethod = CodeGenerationSymbolFactory.CreateMethodSymbol(methodSymbol, name: newMethodIdentifier, parameters: validParameters, statements: newStatements.ToImmutable(), returnType: typeSymbol);
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var newMethodDeclaration = codeGenerationService.CreateMethodDeclaration(newMethod, options: new CodeGenerationOptions(options: options, parseOptions: expression.SyntaxTree.Options));
+            return newMethodDeclaration;
+        }
+
+        /// <summary>
+        /// Generates a method declaration containing a call to the method that introduced the parameter.
+        /// Example:
+        /// 
+        /// ***This is an intermediary step in which the original function has not be updated yet
+        /// public void M(int x, int y)
+        /// {
+        ///     int f = [|x * y|];
+        /// }
+        /// 
+        /// ---------------------------------------------------->
+        /// 
+        /// public void M(int x, int y) // Generated overload
+        /// {
+        ///     M(x, y, x * y);
+        /// }
+        /// 
+        /// public void M(int x, int y) // Original function
+        /// {
+        ///     int f = x * y;
+        /// }
+        /// </summary>
+        private static async Task<SyntaxNode> GenerateNewMethodOverloadAsync(Document document, TExpressionSyntax expression,
+            IMethodSymbol methodSymbol, int insertionIndex, SyntaxGenerator generator, CancellationToken cancellationToken)
+        {
+            var named = methodSymbol.Parameters.Where(parameter => parameter.HasExplicitDefaultValue).Any();
```

Comment:
this isn't the right logic for 'named'.  we need 'named' if we're either mapping to a parametre index that is further than our argument index *or* we have preceding arguments who are both named *and* the name refers to a parameter not at their index.

Reply:
that was removed already but I will need to discuss with you about those cases.